File: backend/app/api/user_defined_fields.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.user_defined_field import (
    UdfEntityType,
    UserDefinedFieldDefinitionCreate,
    UserDefinedFieldDefinitionResponse,
    UserDefinedFieldDefinitionUpdate,
    UserDefinedFieldValueResponse,
    UserDefinedFieldValuesBulkFetch,
    UserDefinedFieldValueUpdate,
)
from app.services import user_defined_field as svc

router = APIRouter(prefix="/user-defined-fields", tags=["user-defined-fields"])


@router.get("/definitions", response_model=list[UserDefinedFieldDefinitionResponse])
async def list_definitions(
    project_id: uuid.UUID,
    entity_type: UdfEntityType,
    db: AsyncSession = Depends(get_db),
) -> list:
    return await svc.list_definitions(db, project_id, entity_type)


@router.post("/definitions", response_model=UserDefinedFieldDefinitionResponse, status_code=201)
async def create_definition(
    data: UserDefinedFieldDefinitionCreate,
    db: AsyncSession = Depends(get_db),
):
    return await svc.create_definition(db, data)


@router.patch("/definitions/{definition_id}", response_model=UserDefinedFieldDefinitionResponse)
async def update_definition(
    definition_id: uuid.UUID,
    data: UserDefinedFieldDefinitionUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Rename and/or change data type. Changing data type clears every
    existing value stored under this field — see
    app/services/user_defined_field.py:update_definition."""
    return await svc.update_definition(db, definition_id, data)


@router.delete("/definitions/{definition_id}", status_code=204)
async def delete_definition(
    definition_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
) -> Response:
    await svc.delete_definition(db, definition_id)
    return Response(status_code=204)


@router.post("/values/bulk-fetch", response_model=list[UserDefinedFieldValueResponse])
async def list_values(
    data: UserDefinedFieldValuesBulkFetch,
    db: AsyncSession = Depends(get_db),
) -> list:
    """Bulk fetch for a grid page — one call for every visible UDF column x
    every visible row."""
    return await svc.list_values_for_records(db, data.field_definition_ids, data.record_ids)


@router.put("/values/{field_definition_id}/{record_id}", response_model=UserDefinedFieldValueResponse)
async def set_value(
    field_definition_id: uuid.UUID,
    record_id: uuid.UUID,
    data: UserDefinedFieldValueUpdate,
    db: AsyncSession = Depends(get_db),
):
    return await svc.set_value(db, field_definition_id, record_id, data)

File: frontend/src/lib/userDefinedFields.ts
import { useCallback, useEffect, useState } from 'react'
import type { IndicatorValue, UdfDataType, UdfEntityType, UserDefinedFieldDefinition, UserDefinedFieldValue } from '@/modules/scheduling/types'
import { api } from './api'

// P6-style User Defined Fields (2026-07-07, per Maro — docs/SCHEDULING_GAPS_PLAN.md
// Phase 9): a project-scoped custom-field definition system (create once per
// project/entity type), plus per-record values, surfaced as an inline-editable
// "(UDF)" column in each entity's own grid. One definitions list per
// (projectId, entityType); values are bulk-fetched separately for whichever
// record ids the host grid currently has on screen.
export function useUserDefinedFieldDefinitions(projectId: string | undefined, entityType: UdfEntityType) {
  const [definitions, setDefinitions] = useState<UserDefinedFieldDefinition[]>([])
  const [loading, setLoading] = useState(true)

  const load = async () => {
    if (!projectId) return
    setLoading(true)
    try {
      const { data } = await api.get<UserDefinedFieldDefinition[]>('/api/v1/user-defined-fields/definitions', {
        params: { project_id: projectId, entity_type: entityType },
      })
      setDefinitions(data)
    } finally {
      setLoading(false)
    }
  }

  useEffect(() => {
    load()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [projectId, entityType])

  const create = async (name: string, dataType: UdfDataType) => {
    await api.post('/api/v1/user-defined-fields/definitions', {
      project_id: projectId, entity_type: entityType, name, data_type: dataType,
    })
    await load()
  }

  const update = async (definitionId: string, changes: { name?: string; data_type?: UdfDataType }) => {
    await api.patch(`/api/v1/user-defined-fields/definitions/${definitionId}`, changes)
    await load()
  }

  const remove = async (definitionId: string) => {
    await api.delete(`/api/v1/user-defined-fields/definitions/${definitionId}`)
    await load()
  }

  return { definitions, loading, create, update, remove, refetch: load }
}

function valueKey(fieldDefinitionId: string, recordId: string): string {
  return `${fieldDefinitionId}:${recordId}`
}

export type UdfValuePayload =
  | { value_text: string }
  | { value_number: string }
  | { value_date: string }
  | { value_indicator: IndicatorValue }
  | { value_text: null; value_number: null; value_date: null; value_indicator: null }

// Bulk-fetches values for a specific set of definitions x records (a grid
// page's own currently-visible rows), keyed for O(1) per-cell lookup —
// call refetch whenever the visible record id set or the definitions
// themselves change.
export function useUserDefinedFieldValues(definitions: UserDefinedFieldDefinition[], recordIds: string[]) {
  const [values, setValues] = useState<Map<string, UserDefinedFieldValue>>(new Map())

  const load = async () => {
    if (definitions.length === 0 || recordIds.length === 0) {
      setValues(new Map())
      return
    }
    const { data } = await api.post<UserDefinedFieldValue[]>('/api/v1/user-defined-fields/values/bulk-fetch', {
      field_definition_ids: definitions.map(d => d.id), record_ids: recordIds,
    })
    setValues(new Map(data.map(v => [valueKey(v.field_definition_id, v.record_id), v])))
  }

  useEffect(() => {
    load()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [definitions.map(d => d.id).join(','), recordIds.join(',')])

  // useCallback, keyed on `values` (2026-07-17 perf fix, per a real bug
  // caught wiring Scheduling.tsx's own "group by a UDF" feature) — this
  // used to be a plain closure, a brand-new function reference every
  // render regardless of whether `values` itself had actually changed.
  // Harmless for a plain per-cell read, but Scheduling.tsx's own grouped-
  // view useMemo needs to depend on getValue to recompute when a UDF value
  // actually changes — an unstable reference there would either force that
  // memo to recompute on literally every render (defeating the whole point
  // of memoizing it) or, if the lint warning were suppressed instead,
  // silently go stale whenever `values` changed without some *other*
  // dependency also changing. Keying the callback on `values` gives a
  // stable reference across renders where nothing here actually changed,
  // and a fresh one exactly when it should.
  const getValue = useCallback((fieldDefinitionId: string, recordId: string): UserDefinedFieldValue | undefined =>
    values.get(valueKey(fieldDefinitionId, recordId)), [values])

  const setValue = async (fieldDefinitionId: string, recordId: string, payload: UdfValuePayload) => {
    const { data } = await api.put<UserDefinedFieldValue>(
      `/api/v1/user-defined-fields/values/${fieldDefinitionId}/${recordId}`, payload
    )
    setValues(prev => new Map(prev).set(valueKey(fieldDefinitionId, recordId), data))
  }

  return { getValue, setValue, refetch: load }
}
